File: training-data/perl/longest_valid_parentheses.pl
use strict;
use warnings;

sub longest_valid {
    my @stack = (-1);
    my $best = 0;
    my @c = split //, shift;
    for my $i (0 .. $#c) {
        if ($c[$i] eq '(') { push @stack, $i; next }
        pop @stack;
        if (@stack) { my $len = $i - $stack[-1]; $best = $len if $len > $best }
        else        { push @stack, $i }
    }
    return $best;
}

printf "%-10s %d\n", "'$_'", longest_valid($_) for '(()', ')()())', '', '()(()', '((()))()';

File: training-data/go/longest_valid_parentheses.go
package main

import "fmt"

func longestValid(s string) int {
	stack := []int{-1}
	best := 0
	for i, c := range s {
		if c == '(' {
			stack = append(stack, i)
			continue
		}
		stack = stack[:len(stack)-1]
		if len(stack) == 0 {
			stack = append(stack, i)
		} else if l := i - stack[len(stack)-1]; l > best {
			best = l
		}
	}
	return best
}

func main() {
	for _, s := range []string{"(()", ")()())", "", "()(()", "((()))()"} {
		fmt.Printf("%q %d\n", s, longestValid(s))
	}
}
